PsfLauncher/PsfPowershellScriptRunner.h

```diff
@@ -0,0 +1,305 @@
+		ThrowIfScriptDoesNotExists(scriptPath);
+
+		bool canScriptRun = false;
+		THROW_IF_FAILED(CheckIfShouldRun(canScriptRun, *startScriptInformation));
+
+		if (!canScriptRun)
+		{
+			return ERROR_SUCCESS;
+		}
+
+
+		bool waitForScriptToFinish = GetWaitForScriptToFinish(*startScriptInformation);
+
+		//Don't run if the users wants the starting script to run asynch
+		//AND to stop running PSF is the starting script encounters an error.
+		//We stop here because we don't want to crash the application if the
+		//starting script encountered an error.
+		if (stopOnScriptError && !waitForScriptToFinish)
+		{
+			THROW_HR_MSG(ERROR_BAD_CONFIGURATION, "PSF does not allow stopping on a script error and running asynchronously.  Please either remove stopOnScriptError or add a wait");
+		}
+
+		std::wstring commandString = MakeCommandString(*startScriptInformation, scriptPath);
+		DWORD scriptTimeout = GetTimeout(*startScriptInformation);
+
+
+		int showWindowAction = GetShowWindowAction(*startScriptInformation);
+		bool runInVirtualEnvironment = GetRunInVirtualEnvironment(*startScriptInformation);
+
+		if (waitForScriptToFinish)
+		{
+			HRESULT startScriptResult = StartProcess(nullptr, commandString.data(), currentDirectory.c_str(), showWindowAction, runInVirtualEnvironment, scriptTimeout);
+
+			if (stopOnScriptError)
+			{
+				THROW_IF_FAILED(startScriptResult);
+			}
+		}
+		else
+		{
+			//We don't want to stop on an error and we want to run async
+			std::thread(StartProcess, nullptr, commandString.data(), currentDirectory.c_str(), showWindowAction, runInVirtualEnvironment, scriptTimeout);
+		}
+
+		return {};
+	}
+
+	void RunEndingScript()
+	{
+		if (!endScriptInformation)
+		{
+			return;
+		}
+
+		const std::wstring& scriptPath = GetScriptPath(*endScriptInformation);
+
+		//This method throws an exception if the script does not exist
+		ThrowIfScriptDoesNotExists(scriptPath);
+
+		bool canScriptRun = false;
+		THROW_IF_FAILED(CheckIfShouldRun(canScriptRun, *endScriptInformation));
+
+		if (canScriptRun)
+		{
+			std::wstring commandString = MakeCommandString(*endScriptInformation, scriptPath);
+			DWORD scriptTimeout = GetTimeout(*endScriptInformation);
+
+
+			int showWindowAction = GetShowWindowAction(*endScriptInformation);
+			bool runInVirtualEnvironment = GetRunInVirtualEnvironment(*endScriptInformation);
+
+			THROW_IF_FAILED(StartProcess(nullptr, commandString.data(), currentDirectory.c_str(), showWindowAction, runInVirtualEnvironment, scriptTimeout));
+		}
+	}
+
+private:
+
+	std::unique_ptr<const psf::json_object> startScriptInformation;
+	std::unique_ptr<const psf::json_object> endScriptInformation;
+	std::filesystem::path currentDirectory;
+	bool stopOnScriptError = false;
+
+	std::wstring MakeCommandString(const psf::json_object& scriptInformation, const std::wstring& scriptPath) noexcept
+	{
+		std::wstring commandString = L"Powershell.exe -file StartingScriptWrapper.ps1 ";
+		commandString.append(L"\".\\");

```

Comment:
>append [](start = 16, length = 6)

Just FYI, I believe append can throw in exceptional cases.

Reply:
Removed noexcept

---
In reply to: [316820139](https://github.com/microsoft/MSIX-PackageSupportFramework/pull/79#discussion_r316820139) [](ancestors = 316820139)